comment logger and danmaku subtitle

tamagoyanki opened the pull request:
Hi, I tried to change **DEFAULTS** -> **"comments"** -> **"record"** to **True** in the file _**settings.py**_ to enable the comment logger. It worked but I think about half of the comments were missing. I would like to make some contributions to it: 

1. Record comments on real time using websocket.
2. Convert comments to a danmaku (弾幕 / bullets) subtitle ass file, which can be watched with the recorded video. The danmaku comments are those flying bullet comments which can be seen on websites like niconico and bilibili.

These are not perfect, and I don't know if you will be interested in watching showroom with comments. Anyway, many thanks to your great work!

tamagoyanki committed:
comment logger and danmaku subtitle

skeith commented:
Is the danmaku style optional?

tamagoyanki commented:
You can change the code for font size, font name, transparency, flight time etc., but there is no optional argument for _**showroom.py**_ or setting in **_settings.py_**.

tamagoyanki committed:
handling exception: WebSocketConnectionClosedException

tamagoyanki committed:
Merge remote-tracking branch 'upstream/master'

tamagoyanki committed:
added logging and quit method
- added cmt_logger for logging
- quit method to stop recording comments and save file

wlerin commented:
Sorry it took me so long to notice this. Let me look over it.

tamagoyanki committed:
Merge remote-tracking branch 'upstream/master'

tamagoyanki commented:
I got a small script for it [[here]](https://github.com/tamagoyanki/sr_danmaku). Maybe you can try it first :)

wlerin commented:
I have one machine running on your branch right now.

It's generating very occasional decoding errors -- looks like for some reason a few messages are encoded in Latin-1 rather than UTF-8. (Guess they were really desperate to save bytes.)

Perhaps as one step in trying to fix broken messages, try decoding the string as Latin-1 before passing it to json.loads

tamagoyanki commented:
There are two kinds of errors so far I noticed. I am not sure which one you found:
1. passed UTF-8 decode but the message is broken and causes json decode error. This one I just pad some dummy message to it, so that it can be decoded by jason.loads. The fixed comments as I observed were all complete sentences, so I think it's OK for now.

2. dose not pass UTF-8 decode and causes the websocket to close. It's even before json.loads. For example there's a viewer id: Kämpfer. Every time when he made comments, the websocket closed. And it does not pass message to the call back function on_message so I am not sure how to deal with it now.

Sorry I am new to python...

wlerin commented:
The second one. It's caused by messages that have only ascii and characters in the Latin-1 extended block, which are all single byte in Latin-1, but the latter require two bytes in UTF-8. Under those conditions the server is encoding, for example, ê as \xea instead of \xc3\xaa, and similar.

Unfortunately, this websocket-client library assumes streams are going to be UTF-8 encoded, so we'd almost have to handle this case in ws_on_error, as the bytes of the message *are* saved in the exception object. 

I can merge this PR and then work on fixing that.

wlerin commented on showroom/comments.py:
```diff
@@ -57,80 +188,275 @@ def __init__(self, room, client, settings, watcher):
+
+        def ws_on_error(ws, error):
+            """ WebSocket callback """
+            cmt_logger.error('websocket on error: {}'.format(error))
```
Fixing the encoding errors might be as simple as changing this to:

```python
        def ws_on_error(ws, error):
            """ WebSocket callback """
            if isinstance(error, UnicodeDecodeError):
                data = error.object
                try:
                    data = data.decode('latin-1')
                except UnicodeDecodeError:
                    pass
                else:
                    ws_on_message(ws, data)
                    return
            cmt_logger.error('websocket on error: {}'.format(error))
```

I don't think Latin-1 decoding can actually throw a UnicodeDecodeError, but there's no performance loss in a try-except that never gets triggered.

tamagoyanki replied:
Thank you for the explanation and the code. I see now we can have the correct encoded message. I have a question that when this method returns, the websocket will still close like before, right? I will make the change but it will take time for me to test it. I cannot record too many rooms on my computer so actually I rarely met this error.

tamagoyanki replied:
If you know what room encountered this error a lot, please let me know.

tamagoyanki committed:
When UnicodeDecodeError happens, decode the message as latin-1

tamagoyanki committed:
change to self.ws.run_forever(skip_utf8_validation=True)

tamagoyanki commented:
I actually changed my id to "¥Kämpfer¥" and threw stars to test. It worked to fix the message and to pass it to _ws_on_message_. It's also needed to change _self.ws.run_forever()_ to _self.ws.run_forever(skip_utf8_validation=True)_ otherwise it throws a different exception rather than UnicodeDecodeError.
The websocket still closes after the fix, but I think I will just try to reconnect it.

tamagoyanki committed:
Reconnect WebSocket if websocket closes due to utf-8 decode error
1. set _needReconnect to True when utf-8 decode error happens
2. after websocket closes, wait for interval thread to finish
3. reconnect if _needReconnect is true
4. reset some variables to their initial values
5. the start time should not be changed

tamagoyanki committed:
try to receive binary data and decode as utf-8 or latin-1
- run_forever() is not used
- now websocket will not close due to UnicodeDecodeError

tamagoyanki commented:
I found a way to receive binary frame data and to decode it as latin-1. Now the websocket will not close due to UnicodeDecodeError and thus no need to reconnect. Now it's needed to handle frame opcode in the lower level like ping, pong, or close opcode, but I didn't see any opcode other than text sent from showroom so maybe we don't need to handle them... I will test to see.

wlerin commented:
Does the websocket close on any error, or only uncaught errors (in this case caused by turning off unicode validation)? If it stayed open on these errors before the change, maybe it would be better to catch the exception it throws instead of UnicodeDecodeError, and avoid having to do this low level stuff manually.

Sorry, I'm very unfamiliar with websockets, which is partly why I never implemented this feature before.

tamagoyanki commented:
It closes either way. When turning on unicode validation, it raises WebSocketPayloadException, sends it to on_error() without original binary data, and then closes connection. When turning off unicode validation, it catches UnicodeDecodeError, sends it to on_error() with original binary data, and then still closes connection.

The commit e029a3e works like this. When there's UnicodeDecodeError, we get original binary data from it and decode it as latin-1. The websocket will close by itself but then we reconnect it within a second. As I observed, in a 3000+ viewer room, it lost at most 3 comments. Usually 0 or only 1 comment was lost. It's not too bad though.

New commit can catch all comments. I just implemented the part for other frame opcodes. I'll let you know how it works.

tamagoyanki committed:
handling other frame opcodes

tamagoyanki committed:
added voting result comment

wlerin commented:
Don't worry about merging the change I just pushed (yet). I need to refine it and make it optional, or maybe better yet put it on a branch for now.

tamagoyanki commented:
I see, thanks.

tamagoyanki committed:
added voting start comment, disabled some debug messages

tamagoyanki commented:
I compared the new version with e029a3e, so far it worked OK and plus solved the UnicodeDecodeError problem. Please let me know if you find any problem, thanks.

wlerin commented:
I haven't seen any errors so far. Maybe there's a few improvements that could still be made, but I think it's ready to merge. If you agree?

tamagoyanki commented:
Yes, it would be great to merge, thank you! BTW just in case you don't have it, [[here]](https://mega.nz/#!EMQxkaYa!TvtYveTrqVX8wwzIPJaLD5Gg--iGF0Y5HlFrNi0bpwE) you can download the font for the subtitles.

tamagoyanki committed:
Merge remote-tracking branch 'upstream/master'
